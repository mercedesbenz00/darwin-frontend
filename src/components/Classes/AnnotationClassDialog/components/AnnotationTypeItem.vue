<template>
  <div
    v-if="annotationType"
    v-tooltip="disabled ? {
      content: disabledTooltip,
      placement: 'left',
      offset: 10
    } : undefined"
    class="annotation-type-item"
    :class="{
      'annotation-type-item--disabled': disabled,
      'annotation-type-item--selected': selected
    }"
    @click="onClick"
  >
    <type-icon
      class="annotation-type-item__icon"
      :type="typeName"
      :color="iconColor"
    />
    <span class="annotation-type-item__label">{{ typeNameFormatted }}</span>
    <!--
      The v-if here is because there is no tooltip defined for the "link" type,
      so any team which has links enabled will get errors otherwise.
      https://linear.app/v7labs/issue/V7-985/add-design-for-link-tooltip-in-annotation-class-dialog
    -->
    <annotation-type-item-info
      v-if="hasInfo"
      class="annotation-type-item__info"
      :data="annotationType"
    />
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import TypeIcon from '@/components/Common/AnnotationType/TypeIcon.vue'
import {
  AnnotationTypeName,
  AnnotationTypePayload,
  FeaturePayload,
  RootState
} from '@/store/types'
import { annotationTypeRichData, AnnotationTypeRichPayload, formatTypeName } from '@/utils'

import AnnotationTypeItemInfo from './AnnotationTypeItemInfo/AnnotationTypeItemInfo.vue'

@Component({
  name: 'annotation-type-item',
  components: { AnnotationTypeItemInfo, TypeIcon }
})
export default class AnnotationTypeItem extends Vue {
  @Prop({ required: true })
  typeName!: AnnotationTypeName

  get typeNameFormatted (): string {
    return formatTypeName(this.typeName)
  }

  @Prop({ type: Boolean, default: false })
  disabled!: boolean

  @Prop({ default: null })
  disabledTooltip!: string | null

  @Prop({ type: Boolean, default: false })
  selected!: boolean

  get iconColor (): string {
    return this.$theme.getColor('colorFeatherLight')
  }

  get hasInfo (): boolean {
    const typesWithNoInfo: AnnotationTypeName[] = ['graph', 'link', 'string', 'table']
    return !typesWithNoInfo.includes(this.typeName)
  }

  onClick (): void {
    if (this.disabled) { return }
    this.$emit('click', this.typeName)
  }

  @State((state: RootState) => state.features.list)
  features!: FeaturePayload[] | null

  @State((state: RootState) => state.aclass.types)
  annotationTypes!: AnnotationTypePayload[]

  /**
   * The following object maps annotation type names to the features
   * name that enables them.
   */
  featureEnabledTypes = {
    graph: 'DOCUMENTS',
    link: 'LINK_TOOL',
    string: 'DOCUMENTS',
    mask: 'RASTERS',
    table: 'DOCUMENTS'
  }

  /**
   * Computed property which is used to figure out with rich annotation types should be included
   * when some annotation types are guarded behind a feature flag.
   *
   * The return type is an object with the hidden annotation type name as key,
   * and the rich annotation type payload to be shown as value.
   */
  get annotationType (): AnnotationTypeRichPayload | null {
    const { annotationTypes, features, typeName } = this

    for (const [featureEnabledTypeName, featureName] of Object.entries(this.featureEnabledTypes)) {
      if (typeName !== featureEnabledTypeName) { continue }

      const feature = features?.find(f => f.name === featureName)
      if (!feature?.enabled) { return null }
    }

    const type = annotationTypes.find(t => t.name === typeName)
    const richType = annotationTypeRichData.find(r => r.name === typeName)

    if (!type || !richType) { return null }

    const richSubs = type.subs.map((sub) => {
      const richSub = annotationTypeRichData.find(r => r.name === typeName) || {}
      if (sub && richSub) {
        return { ...sub, ...richSub }
      }
      return null
    }).filter((sub): sub is AnnotationTypeRichPayload => !!sub)

    if (type && richType) {
      return {
        ...type,
        ...richType,
        richSubs
      }
    }

    return null
  }
}
</script>

<style lang="scss" scoped>
.annotation-type-item {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: 13px;
  padding: 2px 3px;
  border: 1px solid $colorSecondaryLight3;
  transition: background-color .2s;
  cursor: pointer;
  @include noSelect;

  &:hover:not(.annotation-type-item--disabled),
  &:active:not(.annotation-type-item--disabled) {
    border: 1px solid $colorSecondaryLight2;
    background-color: $colorSecondaryLight2;
    transition: none;
  }
}

.annotation-type-item--selected {
  border: 1px solid $colorSecondaryLight;
  background-color: $colorSecondaryLight1;

  &:hover:not(.annotation-type-item--disabled),
  &:active:not(.annotation-type-item--disabled) {
    border: 1px solid $colorSecondaryLight;
    background-color: $colorSecondaryLight1;
  }
}

.annotation-type-item--disabled {
  .annotation-type-item__icon {
    opacity: .3;
  }
  .annotation-type-item__label {
    opacity: .3;
  }
}

.annotation-type-item__icon {
  @include row--center;
  width: 23px;
  height: 23px;
  margin: 0 8px 0 3px;
  object-fit: contain;
}

.annotation-type-item__label {
  flex: 1;
  @include typography(md, default, normal);
  text-align: left;
  color: $colorSecondaryDark;
  white-space: nowrap;
}
</style>
