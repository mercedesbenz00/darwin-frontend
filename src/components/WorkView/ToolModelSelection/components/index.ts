export { default as ClassPill } from './ClassPill.vue'
export { default as ModelIcon } from './ModelIcon.vue'
