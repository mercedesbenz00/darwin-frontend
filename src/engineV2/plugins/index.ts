import attributesPlugin from './attributes'
import boundingBox from './boundingBox'
import boundingBox3d from './boundingBox3d'
import brush from './brush'
import clicker from './click'
import clipboardPlugin from './clipboard'
import commentatorPlugin from './commentator'
import directionalVector from './directionalVector'
import editPlugin from './edit'
import ellipse from './ellipse'
import fieldPlugin from './field'
import inferencePlugin from './inference'
import instanceIDPlugin from './instanceId'
import keypoint from './keypoint'
import link from './link'
import magnifierGlass from './magnifierGlass'
import mask from './mask'
import measuresPlugin from './measures'
import pixelBrush from './pixelBrush'
import polygon from './polygon'
import polyline from './polyline'
import saveAnnotations from './saveAnnotations'
import selectPlugin from './select'
import skeleton from './skeleton'
import tablePlugin from './table'
import tag from './tag'
import textPlugin from './text'
import v7Look from './v7'
import videoPlayerPlugin from './videoPlayer'
import windowLevelPlugin, { WindowLevelPlugin } from './windowLevel'
import zoomPlugin from './zoom'
import zoomCorrection from './zoomCorrection'
export {
  attributesPlugin,
  boundingBox,
  boundingBox3d,
  brush,
  clicker,
  clipboardPlugin,
  commentatorPlugin,
  directionalVector,
  editPlugin,
  ellipse,
  fieldPlugin,
  inferencePlugin,
  instanceIDPlugin,
  keypoint,
  link,
  magnifierGlass,
  mask,
  measuresPlugin,
  pixelBrush,
  polygon,
  polyline,
  saveAnnotations,
  selectPlugin,
  skeleton,
  tablePlugin,
  tag,
  textPlugin,
  v7Look,
  videoPlayerPlugin,
  windowLevelPlugin,
  WindowLevelPlugin,
  zoomPlugin,
  zoomCorrection
}
