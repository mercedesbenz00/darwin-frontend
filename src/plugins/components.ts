import Vue from 'vue'

import NegativeButton from '@/components/Common/Button/V1/NegativeButton.vue'
import PositiveButton from '@/components/Common/Button/V1/PositiveButton.vue'
import PrimaryButton from '@/components/Common/Button/V1/PrimaryButton.vue'
import SecondaryButton from '@/components/Common/Button/V1/SecondaryButton.vue'
import SecondaryLightButton from '@/components/Common/Button/V1/SecondaryLightButton.vue'

export const installCommonComponents = (_Vue: typeof Vue) => {
  _Vue.component('NegativeButton', NegativeButton)
  _Vue.component('PrimaryButton', PrimaryButton)
  _Vue.component('SecondaryButton', SecondaryButton)
  _Vue.component('SecondaryLightButton', SecondaryLightButton)
  _Vue.component('PositiveButton', PositiveButton)
}
