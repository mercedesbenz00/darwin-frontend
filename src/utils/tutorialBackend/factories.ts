import {
  AnnotateStagePayload,
  CodeStagePayload,
  CompleteStagePayload,
  DatasetImagePayload,
  DatasetItemPayload,
  DatasetItemStatus,
  DatasetItemType,
  ImagePayload,
  ModelStagePayload,
  ReviewStagePayload,
  StageType,
  WorkflowStagePayload,
  WorkflowStageTemplatePayload,
  WorkflowTemplatePayload
} from '@/store/types'

import TutorialError from './TutorialError'
import { dataset } from './data/dataset'

const itemId = (datasetImageId: number) => datasetImageId * 10
const workflowId = (datasetImageId: number) => datasetImageId * 100

const stageId = (datasetImageId: number, stageNumber: number) =>
  workflowId(datasetImageId) + stageNumber

const initializeDatasetImage = (image: ImagePayload): DatasetImagePayload => ({
  dataset_id: dataset.id,
  dataset_video_id: null,
  id: image.id * 10,
  image,
  seq: image.id,
  set: 1
})

const initializeItem = (datasetImage: DatasetImagePayload): DatasetItemPayload => ({
  archived_reason: null,
  archived: false,
  current_workflow: null,
  current_workflow_id: null,
  dataset_id: datasetImage.dataset_id,
  dataset_image_id: datasetImage.id,
  dataset_image: datasetImage,
  dataset_video_id: null,
  dataset_video: null,
  file_size: 0,
  filename: datasetImage.image.original_filename,
  height: datasetImage.image.height,
  id: itemId(datasetImage.id),
  inserted_at: '2020-06-12T00:00:00',
  labels: [],
  path: '/',
  priority: 0,
  seq: datasetImage.seq,
  set: datasetImage.set,
  status: DatasetItemStatus.annotate,
  updated_at: '2020-06-12T00:00:00',
  width: datasetImage.image.width,
  type: DatasetItemType.image
})

const initializeStage = (
  datasetImage: DatasetImagePayload,
  stageTemplate: WorkflowStageTemplatePayload
): WorkflowStagePayload => {
  const { type } = stageTemplate
  const payload = {
    assignee_id: null,
    completed: false,
    completes_at: null,
    dataset_item_id: itemId(datasetImage.id),
    id: stageId(datasetImage.id, stageTemplate.stage_number),
    metadata: {},
    number: stageTemplate.stage_number,
    skipped_reason: null,
    skipped: false,
    template_metadata: stageTemplate.metadata,
    type: stageTemplate.type,
    workflow_id: workflowId(datasetImage.id),
    workflow_stage_template_id: stageTemplate.id
  }

  if (type === StageType.Annotate) { return payload as AnnotateStagePayload }
  if (type === StageType.Code) { return payload as CodeStagePayload }
  if (type === StageType.Complete) { return payload as CompleteStagePayload }
  if (type === StageType.Model) { return payload as ModelStagePayload }
  if (type === StageType.Review) { return payload as ReviewStagePayload }

  throw new Error('Invalid type given')
}

const initializeWorkflow = (
  datasetImage: DatasetImagePayload,
  workflowTemplate: WorkflowTemplatePayload,
  currentStage: number
): Pick<DatasetItemPayload, 'current_workflow' | 'current_workflow_id'> => {
  const annotateTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Annotate)

  const reviewTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Review)

  const completeTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Complete)

  if (!annotateTemplate || !reviewTemplate || !completeTemplate) {
    throw new TutorialError('Initializing workflow using invalid template ')
  }

  const id = workflowId(datasetImage.id)

  const workflow: DatasetItemPayload['current_workflow'] = {
    dataset_item_id: itemId(datasetImage.id),
    id,
    workflow_template_id: workflowTemplate.id,
    current_stage_number: currentStage,
    current_workflow_stage_template_id: annotateTemplate.id,
    status: DatasetItemStatus.annotate,
    stages: {
      1: [initializeStage(datasetImage, annotateTemplate)],
      2: [initializeStage(datasetImage, reviewTemplate)],
      3: [initializeStage(datasetImage, completeTemplate)]
    }
  }

  return { current_workflow: workflow, current_workflow_id: id }
}

export const createItem = (
  image: ImagePayload,
  workflowTeplate: WorkflowTemplatePayload,
  stageNumber: number = 1
): DatasetItemPayload => {
  const datasetImage = initializeDatasetImage(image)
  return {
    ...initializeItem(datasetImage),
    ...initializeWorkflow(datasetImage, workflowTeplate, stageNumber)
  }
}
