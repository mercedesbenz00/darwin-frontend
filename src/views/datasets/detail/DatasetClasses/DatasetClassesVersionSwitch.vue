<template>
  <dataset-classes-v2
    v-if="dataset.version === 2"
    :dataset="dataset"
  />
  <dataset-classes
    v-else
    :dataset="dataset"
  />
</template>

<script lang="ts">
import {
  computed,
  defineComponent,
  PropType,
  Ref
} from 'vue'

import { BreadCrumb, useBreadcrumbs } from '@/composables'
import { DatasetPayload } from '@/store/types'
import DatasetClasses from '@/views/datasets/detail/DatasetClasses/V1/DatasetClasses.vue'
import DatasetClassesV2 from '@/views/datasets/detail/DatasetClasses/V2/DatasetClasses.vue'

export default defineComponent({
  name: 'DatasetClassesVersionSwitch',
  components: {
    DatasetClasses,
    DatasetClassesV2
  },
  props: {
    dataset: {
      required: true,
      type: Object as PropType<DatasetPayload>
    }
  },
  setup (props) {
    const breadCrumbs: Ref<BreadCrumb[]> = computed(() => {
      if (props.dataset.version === 2) {
        return [
          { to: '/datasets', name: 'Datasets' },
          { to: `/datasets/${props.dataset.id}/dataset-management`, name: props.dataset.name },
          { name: 'Classes' }
        ]
      } else {
        return [
          { to: '/datasets', name: 'Datasets' },
          { to: `/datasets/${props.dataset.id}/overview`, name: props.dataset.name },
          { to: `/datasets/${props.dataset.id}/classes`, name: 'Classes' }
        ]
      }
    })

    useBreadcrumbs(breadCrumbs)
  }
})
</script>
