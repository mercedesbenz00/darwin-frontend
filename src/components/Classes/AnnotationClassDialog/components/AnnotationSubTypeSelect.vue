<template>
  <annotation-class-section
    v-if="isRendering"
    title="Subtypes"
    optional
  >
    <div class="annotation-types">
      <annotation-type-item
        v-for="type in annotationSubtypes"
        :key="type"
        class="annotation-type-item"
        disabled-tooltip="Cannot remove existing subtypes"
        :type-name="type"
        :disabled="originalSubTypeSelections[type]"
        :selected="typeSelections[type]"
        @click="updateSubAnnotation(type)"
      />
    </div>
  </annotation-class-section>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import Info from '@/components/Common/Info.vue'
import { AnnotationTypeName, AnnotationTypePayload, RootState } from '@/store/types'
import { formatTypeName } from '@/utils'

import AnnotationTypeItem from './AnnotationTypeItem.vue'
import AnnotationClassSection from './Common/AnnotationClassSection.vue'

const ORDERED_SUBS: AnnotationTypeName[] =
  ['attributes', 'text', 'directional_vector', 'instance_id']

@Component({
  name: 'annotation-sub-type-select',
  components: { AnnotationClassSection, AnnotationTypeItem, Info }
})
export default class AnnotationSubTypeSelect extends Vue {
  @Prop({ required: true })
  mainAnnotationType!: AnnotationTypeName | null

  /**
   * Once a subtype has been selected and persisted to backend,
   * it can no longer be removed from a class.
   *
   * This mapping is used to ensure this constraint.
   */
  @Prop({ required: true })
  originalSubTypeSelections!: Partial<Record<AnnotationTypeName, boolean>>

  @Prop({ required: true })
  typeSelections!: Partial<Record<AnnotationTypeName, boolean>>

  @State((state: RootState) => state.aclass.types)
  annotationTypes!: AnnotationTypePayload[]

  get annotationSubtypes (): AnnotationTypeName[] {
    const type =
      this.annotationTypes.find(t => t.name === this.mainAnnotationType && t.granularity === 'main')

    if (!type) {
      throw new Error('Invalid main type given as prop')
    }

    const typeSubs = type.subs.map(s => s.name)
    return ORDERED_SUBS.filter(s => typeSubs.includes(s))
  }

  get isRendering (): boolean {
    return this.annotationSubtypes.length > 0
  }

  formatTypeName = formatTypeName

  updateSubAnnotation (typeName: AnnotationTypeName) {
    this.$emit('update:typeSelections', {
      ...this.typeSelections,
      [typeName]: !this.typeSelections[typeName]
    })
  }
}
</script>

<style lang="scss" scoped>
.annotation-types {
  display: block;
  width: 100%;
}

.annotation-type-item {
  width: calc(50% - 10px);
  margin: 5px 0;

  &:nth-child(even) {
    margin-left: 10px;
  }

  &:nth-child(odd) {
    margin-right: 10px;
  }
}
</style>
