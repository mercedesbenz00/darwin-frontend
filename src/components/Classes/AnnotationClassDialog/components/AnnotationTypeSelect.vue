<template>
  <annotation-class-section
    ref="container"
    title="Annotation Type"
    :error="error"
  >
    <div class="annotation-types">
      <annotation-type-item
        v-for="type in mainAnnotationTypeNames"
        :key="type"
        class="annotation-type-item"
        disabled-tooltip="Cannot edit main annotation type"
        :type-name="type"
        :disabled="editing"
        :selected="selectedAnnotationType === type"
        @click="selectAnnotation(type)"
      />
    </div>
  </annotation-class-section>
</template>

<script lang="ts">
import { startCase } from 'lodash'
import { Component, Prop, Vue } from 'vue-property-decorator'
import { Getter } from 'vuex-class'

import Info from '@/components/Common/Info.vue'
import { AnnotationTypeName, AnnotationTypePayload } from '@/store/types'

import AnnotationTypeItem from './AnnotationTypeItem.vue'
import AnnotationClassSection from './Common/AnnotationClassSection.vue'

const ORDERED_TYPES: AnnotationTypeName[] = [
  'polygon',
  'bounding_box',
  'tag',
  'mask',
  'skeleton',
  'line',
  'keypoint',
  'ellipse',
  'cuboid',
  'link',
  'string',
  'graph',
  'table'
]

@Component({
  name: 'annotation-type-select',
  components: { AnnotationClassSection, AnnotationTypeItem, Info }
})
export default class AnnotationTypeSelect extends Vue {
  @Getter('mainAnnotationTypes', { namespace: 'aclass' })
  mainAnnotationTypes!: AnnotationTypePayload[]

  @Prop({ default: null, required: true })
  selectedAnnotationType!: AnnotationTypeName | null

  @Prop({ type: Boolean, default: false, required: false })
  editing!: boolean

  @Prop({ default: null, required: false })
  error!: string | null

  $refs!: {
    container: AnnotationClassSection
  }

  get mainAnnotationTypeNames (): AnnotationTypeName[] {
    const typeMains = this.mainAnnotationTypes.map(s => s.name)
    return ORDERED_TYPES.filter(s => typeMains.includes(s))
  }

  /**
   * Annotation type names are in `snake_case`.
   *
   * This function converts them to a format frendlier for display by
   * replacing underscores with spaces and capitalizing the first letter of
   * each word.
   *
   * This is called `startCase` in lodash
   */
  formatTypeName = startCase

  public scrollTo (): void {
    this.$refs.container.scrollTo()
  }

  selectAnnotation (typeName: AnnotationTypeName): void {
    if (this.error) {
      this.$emit('update:error', null)
    }
    this.$emit('update:selectedAnnotationType', typeName)
  }
}
</script>

<style lang="scss" scoped>
.annotation-types {
  display: block;
  width: 100%;
}

.annotation-type-item {
  width: calc(50% - 10px);
  margin: 5px 0;

  &:nth-child(even) {
    margin-left: 10px;
  }

  &:nth-child(odd) {
    margin-right: 10px;
  }
}
</style>
