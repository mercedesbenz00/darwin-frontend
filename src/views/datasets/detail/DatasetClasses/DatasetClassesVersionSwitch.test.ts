import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { buildDatasetPayload } from 'test/unit/factories'

import { installCommonComponents } from '@/plugins/components'
import DatasetClassesVersionSwitch from '@/views/datasets/detail/DatasetClasses/DatasetClassesVersionSwitch.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
installCommonComponents(localVue)

let store: ReturnType<typeof createTestStore>

const sfhDataset = buildDatasetPayload({
  id: 1,
  name: 'Dataset'
})

beforeEach(() => { store = createTestStore() })

it('matches snapshot for v1.0 dataset', () => {
  const propsData = { dataset: sfhDataset }
  const wrapper = shallowMount(DatasetClassesVersionSwitch, { localVue, propsData, store })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot for v2.0 dataset', () => {
  const propsData = { dataset: { ...sfhDataset, version: 2 } }
  const wrapper = shallowMount(DatasetClassesVersionSwitch, { localVue, propsData, store })
  expect(wrapper).toMatchSnapshot()
})
