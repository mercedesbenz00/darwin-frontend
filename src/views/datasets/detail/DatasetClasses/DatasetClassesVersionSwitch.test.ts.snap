// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot for v1.0 dataset 1`] = `<dataset-classes-stub dataset="[object Object]"></dataset-classes-stub>`;

exports[`matches snapshot for v2.0 dataset 1`] = `<dataset-classes-v2-stub dataset="[object Object]"></dataset-classes-v2-stub>`;
